sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -59,170 +56,130 @@ class EthService(val stateView: AccountStateView, val nodeView: CurrentView[Acco
-    val ethTx = new ListBuffer[EthereumTransaction]
-    block.transactions.foreach(transaction =>
-      if (transaction.isInstanceOf[EthereumTransaction]) ethTx.append(transaction.asInstanceOf[EthereumTransaction]))
-    ethTx
+    val transactions = block.get().transactions.filter {
+      _.isInstanceOf[EthereumTransaction]
+    } map {
+      _.asInstanceOf[EthereumTransaction]
+    }
+    new EthereumBlock(
+      Numeric.prependHexPrefix(Integer.toHexString(nodeView.history.getBlockHeight(blockId).get())),
+      Numeric.prependHexPrefix(blockId),
+      if (!hydratedTx) {
+        transactions.map(tx => Numeric.prependHexPrefix(tx.id)).toList.asJava
+      } else {
+        transactions.flatMap(tx => stateView.getTransactionReceipt(Numeric.hexStringToByteArray(tx.id)) match {
+          case Some(receipt) => Some(new EthereumTransactionView(receipt, tx))
+          case None => None
+        }).toList.asJava
+      },
+      block.get())
+  }
+
+  private def doCall(params: TransactionArgs, tag: Quantity) = {
+    getStateViewAtTag(tag) {
+      case None => throw new IllegalArgumentException(s"Unable to get state for given tag: ${tag.getValue}")
+      case Some(tagStateView) => tagStateView.applyMessage(params.toMessage) match {
+          case Some(success: ExecutionSucceeded) => success
+
+          case Some(failed: ExecutionFailed) =>
+            // throw on execution errors, also include evm revert reason if possible
+            throw new RpcException(new RpcError(
+              RpcCode.ExecutionError.getCode, failed.getReason.getMessage, failed.getReason match {
+                case evmRevert: EvmException => Numeric.toHexString(evmRevert.returnData)
+                case _ => null
+              }))
+
+          case Some(invalid: InvalidMessage) =>
+            throw new IllegalArgumentException("Invalid message.", invalid.getReason)
+
+          case _ => throw new IllegalArgumentException("Unable to process call.")
+        }
+    }
   }
 
   @RpcMethod("eth_call") def call(params: TransactionArgs, tag: Quantity): String = {
-    using(getStateViewFromBlockById(getBlockIdByTag(tag.getValue))) { stateDbRoot =>
-      if (stateDbRoot == null)
-        return null
-      else {
-        val result = Evm.Apply(
-          stateDbRoot.getStateDbHandle,
-          params.getFrom,
-          if (params.to == null) null else params.to.toBytes,
-          params.value,
-          params.getData,
-          params.gas,
-          params.gasPrice
-        )
-
-        if (result.evmError.nonEmpty) {
-          throw new RpcException(
-            new RpcError(
-              RpcCode.ExecutionError.getCode,
-              result.evmError,
-              Numeric.toHexString(result.returnData)
-            )
-          )
-        }
-        if (result.returnData == null)
-          null
-        else
-          Numeric.toHexString(result.returnData)
+    val result = doCall(params, tag)
```

Comment:
why we don't catch the potential exceptions here?
I guess it can lead to the internal server error (500) result.

Reply:
The calling `RpcHandler` will catch any exceptions and handle them accordingly:
- If it is an `RpcException` it will unwrap and return the contained `RpcError` response.
- Any other exception will be returned as an `RpcError` with the code of "internal error" and the message of the exception.